runtime/minor_gc.c

```diff
@@ -477,15 +477,32 @@ CAMLexport void caml_gc_dispatch (void)
    the gc. */
 void caml_alloc_small_dispatch (tag_t tag, intnat wosize, int flags)
 {
-  /* Async callbacks may fill the minor heap again, so we need a while
-     loop here. */
-  while (Caml_state->young_ptr < Caml_state->young_trigger){
-    Caml_state->young_ptr += Whsize_wosize (wosize);
+  intnat whsize = Whsize_wosize (wosize);
+
+  /* First, we un-do the allocation performed in [Alloc_small] */
+  Caml_state->young_ptr += whsize;
+
+  while(1) {
+    /* We might be here because of an async callback / urgent GC
+       request. Take the opportunity to to what needs to be done. */
+    if (flags & CAML_FROM_CAML) caml_check_urgent_gc (Val_unit);
+    else caml_check_gc_without_async_callbacks(Val_unit);
```

Comment:
In bytecode mode, we need to call `caml_gc_dispatch` last, because we need the guarantee that there is enough space on the minor heap. So let's leave this as it is in this PR.

Reply:
The two behaviors will be made the same with #8805.